analyses/mutational-signatures/02-explore_de_novo.Rmd

```diff
@@ -0,0 +1,97 @@
+---
+title: "Explore de novo mutational signatures"
+author: Anna R. Poetsch; adapted by J. Taroni for ALSF CCDL
+date: 2020
+output:
+  html_notebook:
+    toc: yes
+    toc_float: yes
+params:
+  is_ci: FALSE
+---
+
+## Determine Signatures de novo  
+
+This notebook is included, because the previous strategies imply that there are mutagenic mechanisms in operation that are unlikely to be a major factor in this dataset, such as smoking and UV light.
+This is happening due to overfitting problems when fitting existing signatures. 
+De novo signature calling is used to provide an alternative for discussion. 
+It is determined, which signatures can be extracted from the data themselves. 
+These signatures are then fitted back to the known signatures that are used for the fitting approach above. 
+
+The signatures were extracted with `scripts/de_novo_signature_fitting.R`.
+You can see the parameters we used in `run_mutational_signatures.sh`.
+
+```{r}
+library(deconstructSigs)
+library(sigfit)
+
+# Assigning params$is_ci to running_in_ci avoids a locked binding error
+running_in_ci <- params$is_ci
+
+# Are we testing? In case of a non 0/1 number, we recast as logical, 
+# and then ensure logical.
+if (running_in_ci %in% c(0,1)) running_in_ci <- as.logical(running_in_ci)
+if (!(is.logical(running_in_ci))) {
+  stop("\n\nERROR: The parameter `is_ci` should be FALSE/TRUE (or 0/1).")
+}
+
+```
+
+```{r}
+# Read in all the results from the range of k as chosen in 
+# run_mutational_signatures.sh
+sigfit_results <- readr::read_rds(file.path("results", 
+                                            "denovo_sigfit_signatures.RDS"))
+```
+
+TODO: Add some analyses in support of picking `nsignatures=10` ?
```

Comment:
Thank you very much for integrating this!
The Plot that shows the fitting of the signatures is the most helpful. Usually also the Frobenius reconstruction error is used for deciding which signatures are most suitable, together with a measure of reproducibility of the signatures. 
However, all these measures do not really give clear cut-offs, which is why deciding on the total numbers is still a bit of a handwaving affair. I usually look at the plot I have included in the code and then look at the signatures that are created in the range where additional signatures do not improve the fitting and then decide by eye, where the cut-off should be, based on whether the additional signatures make biological sense to be separated out (when going to far, it is usually clear, because additional signatures look very alike another signature). 
To come back to your question:
- goodness of Fit (example attached)
- possibly the Frobenius reconstruction error (example attached, but not implemented in the code)
- possibly the signature reproducibility (example attached, but not implemented in the code)
- the profiles of the signatures 
- cosine similarity with known signatures 

![image](https://user-images.githubusercontent.com/34540437/95737453-fbd39d80-0c87-11eb-890f-fadd3e34f0b7.png)
![image](https://user-images.githubusercontent.com/34540437/95737495-10179a80-0c88-11eb-9e5b-981a35b02235.png)

Reply:
@arpoe do you happen to have code for

> * possibly the Frobenius reconstruction error (example attached, but not implemented in the code)
> * possibly the signature reproducibility (example attached, but not implemented in the code)

that I can use as a reference for implementation as part of this project?

Reply:
I was dreading this question... In the sigfit package these measures are not implemented, so to include it, it would have to be implemented from scratch, which I did not do at the time. They are implemented with other packages that have other drawbacks.
To implement these measures for sigfit is on my list of things to do with my prospective students, but we are not quite there, yet.

Reply:
No problem, that's what I thought might be the case based on searching "sigfit Frobenius reconstruction error" a bunch of different ways 🙂   Seems like that effort is beyond the scope of this particular project. Regarding your point above:

> based on whether the additional signatures make biological sense to be separated out (when going too far, it is usually clear, because additional signatures look very alike another signature).

Do you usually look at the mutational spectra to make this determination? Or is there something else I should be doing?

Reply:
Yes, I have to admit that deciding on the right number is really not a clear cut strategy. It is based on the plot for the goodness of fit and then I really just look at the signatures that are generated and their contributions to different samples. 
It is noticeable, when signatures split up without this being very biologically meaningful.
It is also a good measure, if signatures are similar to previously described ones. 
If signatures always go hand in hand for their contribution, it is also a sign that they may be better considered as one. 
Taking all these things into consideration I make the decision on the precise number relative to the numbers indicated by the goodness of fit itself. For the latter I do however not consider the red dot (that is shown in the example above), but rather the numbers just before the graph plateaus. For the example above they are the same, but most of the time, the red dot is too conservative.
I am sorry that this is all a bit too imprecise. This is also something we will work on (the student starts in November).

Reply:
All of this sounds good. I am fine with picking a less conservative number of signatures than the goodness-of-fit plot suggests. I mostly want to make sure to include the relevant information in the notebook that people would typically use to make decisions.